src/lib/ptree/ptree.lua

```diff
@@ -602,55 +594,31 @@ end
 function Manager:push_notifications_to_peers()
    local notifications = alarms.notifications()
    if #notifications == 0 then return end
-   local function head (queue)
-      local msg = assert(queue[1])
-      local len = #msg
-      return ffi.cast('uint8_t*', msg), len
-   end
-   local function tojson (output, str)
-      json_lib.write_json_object(output, str)
-      local msg = output:flush()
+   local function tojson (str)
+      local msg = mem.call_with_output_string(function(output)
+         json_lib.write_json_object(output, str)
+         return output:flush()
+      end)
       return tostring(#msg)..'\n'..msg
    end
    -- Enqueue notifications into each peer queue.
    local peers = self.notification_peers
    for _,peer in ipairs(peers) do
-      local output = json_lib.buffered_output()
       peer.queue = peer.queue or {}
       for _,each in ipairs(notifications) do
-         table.insert(peer.queue, tojson(output, each))
+         table.insert(peer.queue, tojson(each))
       end
    end
    -- Iterate peers and send enqueued messages.
-   for i,peer in ipairs(peers) do
-      local queue = peer.queue
-      while #queue > 0 do
-         local buf, len = head(peer.queue)
-         peer.pos = peer.pos or 0
-         local count, err = peer.fd:write(buf + peer.pos,
-                                          len - peer.pos)
-         if not count then
-            if err.AGAIN then break end
-            peer.state = 'error'
-            peer.msg = tostring(err)
-         elseif count == 0 then
-            peer.state = 'error'
-            peer.msg = 'short write'
-         else
-            peer.pos = peer.pos + count
-            assert(peer.pos <= len)
-            if peer.pos == len then
-               peer.pos = 0
-               table.remove(peer.queue, 1)
-            end
-         end
-
-         if peer.state == 'error' then
-            if peer.state == 'error' then self:warn('%s', peer.msg) end
-            peer.fd:close()
-            table.remove(peers, i)
-         end
-      end
+   for _, peer in ipairs(peers) do
+      fiber.spawn(function()
+         self:call_with_cleanup(peer, function()
+            local msg = peer.queue[1]
+            peer:write_chars(msg)
+            peer:flush_output()
+            table.remove(peer.queue, 1)
```

Comment:
I think you don't want to spawn a fiber here.  With fibers, generally what you want to do is to spawn one fiber per peer:

```lua
function Manager:send_notifications_to_peer(peer)
   self:call_with_cleanup(peer, function()
      while true do
         local msg = notification_queue:get()
         peer:write_chars(msg)
         peer:flush_output()
      end
   end)
end
```

i.e. the fiber lives as long as the peer is connected.  So what about the queue?  The peculiar thing is that it needs to cause the fiber to block when the queue is empty, and only wake up the fiber when something is added to the queue.  For that I would usually recommend a concurrent ML queue; see `lib.fibers.queue`.  A queue in CML is really just a pair of channels, with a dedicated fiber that does the queue functionality.

In your code, you were spawning a fiber per notification.  In the case where writing the notification takes time, you could have two fibers writing on the same socket!  Additionally after writing, the first finishing fiber closed the socket -- that's probably where your EOF comes from.

Regarding queues, I do have a couple of doubts.  One is that really you can only push onto a queue from a fiber.  So you'd need to have a fiber polling for alarms.notifications(), and then putting items on each peer's queue.  That's fine:

```
-- spawn in a fiber
function Manager:notification_poller()
   while true do
      if #alarms.notifications == 0 then
         fiber.sleep(1/50) -- poll at 50 Hz
      else
         -- here we enqueue the notifications to peer fibers
      end
   end
end
```

Then you remove `push_notifications_to_peers`, as running the scheduler will run the polling fiber.  That's nice.

However my doubt is about the queue.  The way fibers works currently is that when a fiber causes another fiber to be runnable, that fiber is enqueued to run on the next turn.  Similarly, if a fiber blocks, the soonest it could run again would be the next turn.  So when a queue is empty and you push a message to it, you wake up the queue fiber and you continue.  If you try to push another message, you will block, because the queue's fiber isn't ready to receive another value yet -- it needs to run first.  Even though the queue is asynchronous, it takes turns to communicate values.  So if you have a lot of notifications coming through, their dispatch rate could be limited by the breath rate, if each breath runs one turn.

There are various ways we could solve this.  We could resume fibers into the runqueue of the current turn, but that could lead to the process devoting proportionally too much time to fibers, when maybe it has some other things it needs to do as well.  Not sure.  It could be of course that this isn't a problem at all!